src/QuadGK.jl

```diff
@@ -91,45 +91,28 @@ function evalrule(f, a,b, x,w,gw, nrm)
 end
 
 rulekey(::Type{BigFloat}, n) = (BigFloat, precision(BigFloat), n)
-rulekey(T,n) = (T,n)
+rulekey(::Type{T},n) where {T<:AbstractFloat} = (T,precision(T),n)
+
+
+function do_quadgk(f, s::TS, n, ::Type{Tw}, abstol, reltol, maxevals, nrm) where {Tw,R<:Real,TS<:AbstractArray{R}}
+    preprocess_real(f,s,n,Tw,abstol, reltol, maxevals, nrm)
+end
+
+function do_quadgk(f, s::TS, n, ::Type{Tw}, abstol, reltol, maxevals, nrm) where {Tw,TS<:AbstractArray}
+    do_quadgk_common(f, s, n, Tw, abstol, reltol, maxevals, nrm)
+end
 
 # Internal routine: integrate f over the union of the open intervals
 # (s[1],s[2]), (s[2],s[3]), ..., (s[end-1],s[end]), using h-adaptive
 # integration with the order-n Kronrod rule and weights of type Tw,
 # with absolute tolerance abstol and relative tolerance reltol,
 # with maxevals an approximate maximum number of f evaluations.
-function do_quadgk(f, s, n, ::Type{Tw}, abstol, reltol, maxevals, nrm) where Tw
-    if eltype(s) <: Real # check for infinite or semi-infinite intervals
-        s1 = s[1]; s2 = s[end]; inf1 = isinf(s1); inf2 = isinf(s2)
-        if inf1 || inf2
-            if inf1 && inf2 # x = t/(1-t^2) coordinate transformation
-                return do_quadgk(t -> begin t2 = t*t; den = 1 / (1 - t2);
-                                            f(t*den) * (1+t2)*den*den; end,
-                                 map(x -> isinf(x) ? copysign(one(x), x) : 2x / (1+hypot(1,2x)), s),
-                                 n, Tw, abstol, reltol, maxevals, nrm)
-            end
-            s0,si = inf1 ? (s2,s1) : (s1,s2)
-            if si < 0 # x = s0 - t/(1-t)
-                return do_quadgk(t -> begin den = 1 / (1 - t);
-                                            f(s0 - t*den) * den*den; end,
-                                 reverse!(map(x -> 1 / (1 + 1 / (s0 - x)), s)),
-                                 n, Tw, abstol, reltol, maxevals, nrm)
-            else # x = s0 + t/(1-t)
-                return do_quadgk(t -> begin den = 1 / (1 - t);
-                                            f(s0 + t*den) * den*den; end,
-                                 map(x -> 1 / (1 + 1 / (x - s0)), s),
-                                 n, Tw, abstol, reltol, maxevals, nrm)
-            end
-        end
-    end
+function do_quadgk_common(f, s::TS, n, ::Type{Tw}, abstol, reltol, maxevals, nrm) where {Tw,TS<:AbstractArray}
 
     key = rulekey(Tw,n)
     x,w,gw = haskey(rulecache, key) ? rulecache[key] :
        (rulecache[key] = kronrod(Tw, n))
-    segs = Segment[]
-    for i in 1:length(s) - 1
-        heappush!(segs, evalrule(f, s[i],s[i+1], x,w,gw, nrm), Reverse)
-    end
+    segs = sort([evalrule(f, s[i], s[i+1], x, w, gw, nrm) for i in 1:length(s)-1], order = Reverse)
```

Comment:
shouldn't this be `heapify!` and not `sort`?

Reply:
fixed